File: ueueo-framework/ueueo-authorization/src/main/java/com/ueueo/authorization/permissions/PermissionSimpleStateCheckerExtensions.cs
﻿using JetBrains.Annotations;
using Volo.Abp.SimpleStateChecking;

namespace Volo.Abp.Authorization.Permissions;

public static class PermissionSimpleStateCheckerExtensions
{
    public static TState RequireAuthenticated<TState>(@Nonnull this TState state)
        //where TState : IHasSimpleStateCheckers<TState>
    {
        state.StateCheckers.Add(new RequireAuthenticatedSimpleStateChecker<TState>());
        return state;
    }

    public static TState RequirePermissions<TState>(
        @Nonnull this TState state,
        params String[] permissions)
        //where TState : IHasSimpleStateCheckers<TState>
    {
        state.RequirePermissions(requiresAll: true, batchCheck: true, permissions);
        return state;
    }

    public static TState RequirePermissions<TState>(
        @Nonnull this TState state,
        boolean requiresAll,
        params String[] permissions)
        //where TState : IHasSimpleStateCheckers<TState>
    {
        state.RequirePermissions(requiresAll: requiresAll, batchCheck: true, permissions);
        return state;
    }

    public static TState RequirePermissions<TState>(
        @Nonnull this TState state,
        boolean requiresAll,
        boolean batchCheck,
        params String[] permissions)
        //where TState : IHasSimpleStateCheckers<TState>
    {
        Objects.requireNonNull(state, nameof(state));
        Check.NotNullOrEmpty(permissions, nameof(permissions));

        if (batchCheck)
        {
            RequirePermissionsSimpleBatchStateChecker<TState>.Current.AddCheckModels(new RequirePermissionsSimpleBatchStateCheckerModel<TState>(state, permissions, requiresAll));
            state.StateCheckers.Add(RequirePermissionsSimpleBatchStateChecker<TState>.Current);
        }
        else
        {
            state.StateCheckers.Add(new RequirePermissionsSimpleStateChecker<TState>(new RequirePermissionsSimpleBatchStateCheckerModel<TState>(state, permissions, requiresAll)));
        }

        return state;
    }
}

File: ueueo-frameworks/ueueo-authorization/src/main/java/com/ueueo/authorization/permissions/PermissionSimpleStateCheckerExtensions.java
package com.ueueo.authorization.permissions;

import com.ueueo.simplestatechecking.IHasSimpleStateCheckers;
import com.ueueo.users.ICurrentUser;
import org.springframework.lang.NonNull;
import org.springframework.util.Assert;

import java.util.List;

public class PermissionSimpleStateCheckerExtensions {

    public static <TState extends IHasSimpleStateCheckers<TState>> TState requireAuthenticated(@NonNull TState state, ICurrentUser currentUser)
    {
        state.getStateCheckers().add(new RequireAuthenticatedSimpleStateChecker<>(currentUser));
        return state;
    }

    public static <TState extends IHasSimpleStateCheckers<TState>> TState requirePermissions(
            @NonNull TState state,
            List<String> permissions,
            IPermissionChecker permissionChecker) {
        requirePermissions(state, true, true, permissions,permissionChecker);
        return state;
    }

    public static <TState extends IHasSimpleStateCheckers<TState>> TState requirePermissions(
            @NonNull TState state,
            boolean requiresAll,
            List<String> permissions,
            IPermissionChecker permissionChecker) {
        requirePermissions(state, requiresAll, true, permissions,permissionChecker);
        return state;
    }

    public static <TState extends IHasSimpleStateCheckers<TState>> TState requirePermissions(
            @NonNull TState state,
            boolean requiresAll,
            boolean batchCheck,
            List<String> permissions,
            IPermissionChecker permissionChecker) {
        Assert.notNull(state, "state must not null!");
        Assert.notEmpty(permissions, "permissions must not empty!");

        if (batchCheck) {
            //TODO 使用线程缓存，不知道为什么
//            RequirePermissionsSimpleBatchStateChecker<TState>.Current.AddCheckModels (new RequirePermissionsSimpleBatchStateCheckerModel<TState>(state, permissions, requiresAll));
//            state.getStateCheckers().add( RequirePermissionsSimpleBatchStateChecker<TState>.Current);
            RequirePermissionsSimpleBatchStateChecker<TState> batchStateChecker = new RequirePermissionsSimpleBatchStateChecker<>(permissionChecker);
            batchStateChecker.addCheckModels(new RequirePermissionsSimpleBatchStateCheckerModel<>(state, permissions, requiresAll));
            state.getStateCheckers().add(batchStateChecker);
        } else {
            state.getStateCheckers().add(new RequirePermissionsSimpleStateChecker<>(permissionChecker,new RequirePermissionsSimpleBatchStateCheckerModel<TState>(state, permissions, requiresAll)));
        }

        return state;
    }
}
